### PR TITLE
[Forkless SparseML Transformers] [Feature Branch] Modding `Mistral`

### PR DESCRIPTION
**Pull request by dbogunowicz**

### CODE REVIEW

File: src/sparseml/transformers/sparsification/trainer.py [hunk 1]
```diff
@@ -894,8 +894,10 @@ def save_optimizer_and_scheduler(self, output_dir: Optional[str] = None):
         if output_dir is None:
             output_dir = self.args.output_dir
 
-        if self.sharded_ddp == ShardedDDPOption.SIMPLE and self.optimizer is not None:
-            self.optimizer.consolidate_state_dict()
+        # TODO: The forkless transformers dependency does not
+        # support ShardedDDPOption anymore. To be investigated.
```
Comment (mgoin) [inline src/sparseml/transformers/sparsification/trainer.py:898]: i think this was simply removed in the more modern versions of transformers

File: tests/sparseml/transformers/sparsification/modification/test_modifying_mistral.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+from accelerate import init_empty_weights
+from sparseml.transformers.sparsification.modification import modify_model
+
+
+@pytest.fixture
+def mistral_model():
+    config = AutoConfig.from_pretrained("mistralai/Mistral-7B-v0.1")
+    with init_empty_weights():
+        model = AutoModel.from_config(config)
+    return model
```
Comment (mgoin) [inline tests/sparseml/transformers/sparsification/modification/test_modifying_mistral.py:29]: nice! so not much memory gets allocated but we get to try a "real size" model?

Comment (dbogunowicz) [reply tests/sparseml/transformers/sparsification/modification/test_modifying_mistral.py:29]: did not investigate the memory requirement, but this model gets initialized in a fraction of a second :)

File: tests/sparseml/transformers/sparsification/modification/test_modifying_mistral.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    mistral_ = deepcopy(mistral_model)
+    mistral = modify_model(mistral_model)
+
+    # check how many modified "MistralAttention" modules are in the original
+    # model (should be 0, as the model is not modified yet)
+    modified_modules_original_model = [
+        module
+        for module in mistral_.modules()
+        if _is_mistral_attention_modified(module)
+        and module.__class__.__name__ == "MistralAttention"
```
Comment (bfineran) [inline tests/sparseml/transformers/sparsification/modification/test_modifying_mistral.py:47]: same here, let's use the class reference itself

**Pull request merged**